### PR TITLE
Initial Commit of sendBeacon Tests

### PR DESCRIPTION
**Pull request by Brandr0id**

### CODE REVIEW

**Comment by Brandr0id** [comment]
@toddreifsteck @igrigorik

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258726553)

# Firefox (nightly) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258726554)

# Sauce (safari) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258726560)

# Chrome (unstable) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/258726561)

# Sauce (MicrosoftEdge) #

File: beacon/beacon-common.js
Comment (igrigorik) [inline beacon/beacon-common.js]: 👍

File: beacon/beacon-redirect.window.js
Comment (igrigorik) [inline beacon/beacon-redirect.window.js]: Awesome 👍

**Review by igrigorik — APPROVED** [review]
Looks great, kudos for the thorough work Brandon!

**Comment by igrigorik** [comment]
@siusin @jgraham any idea on why the bots are failing here and not reporting the results? One of the builds appears to be [erroring out with a Selenium error](https://travis-ci.org/w3c/web-platform-tests/jobs/258726554), but I don't see how or why that would be caused by the code here.

In the meantime, I ran the tests locally on my machine and the results make sense -- e.g. Blob tests fail on Chrome because those are currently disabled.

File: beacon/beacon-cors.window.js
Comment (igrigorik) [inline beacon/beacon-cors.window.js]: > Beacons use the HTTP POST method, which is a CORS-safelisted method, and thus they do not trigger preflight.

This is not true, the behavior here is also conditional on the type of data being transferred. Specifically, "no-cors" applies for CORS-safelisted content-types (`application/x-www-form-urlencoded`, `multipart/form-data`, or `text/plain`), other types change cors mode to "cors" and require a preflight.

Processing section covers this: https://w3c.github.io/beacon/#sec-processing-model

Could we extend these tests to account for the content-type differences?

**Review by igrigorik — CHANGES_REQUESTED** [review]
On second pass, a few CORS-related questions before we land this..

File: beacon/beacon-basic-string.html
Comment (cdumez) [inline beacon/beacon-basic-string.html]: FYI, the LargeString check is slow is causes the test to flakily timeout for WebKit.

Comment (igrigorik) [inline beacon/beacon-basic-string.html]: The string is large, but not that large.. Do you have any suggestions for how to improve this?

**Comment by igrigorik** [comment]
/me nudges @siusin @jgraham, any ideas for: https://github.com/w3c/web-platform-tests/pull/6655#issuecomment-319775273 -- thanks :)

**Comment by siusin** [comment]
pinged @sideshowbarker 

Mike said this Safari failure is a known issue #6602 , he will look at the Chrome failure later.

File: beacon/beacon-basic-string.html
Comment (cdumez) [inline beacon/beacon-basic-string.html]: I'll will investigate on WebKit side. The test takes up to ~130 seconds on some of our debug bots.
In any case, feel free to land as is. It could simply be an issue with our implementation as it is very new.

**Comment by sideshowbarker** [comment]
The Chrome failure is not due to flakiness but instead just due to the build taking longer than the limit that Travis sets for builds to complete. So as soon as all review comments are resolved and a reviewer approves the PR, I can merge it

**Comment by igrigorik** [comment]
@sideshowbarker interesting, presumably the long build is not due to the tests introduced here? Are you seeing similar behavior on other PRs?

@Brandr0id  do you have bandwidth to iterate on CORS issue we flagged above? https://github.com/w3c/web-platform-tests/pull/6655#discussion_r130978823. If not, we can land this and take that as a separate PR.. I don't want this PR to bitrot.

**Comment by sideshowbarker** [comment]
> @sideshowbarker interesting, presumably the long build is not due to the tests introduced here? Are you seeing similar behavior on other PRs?

Actually I misspoke—I restarted the job and it now is (sorta) indicating unstable results:

https://github.com/w3c/web-platform-tests/pull/6655#issuecomment-318779277
https://travis-ci.org/w3c/web-platform-tests/jobs/258726560

But those results themselves are weird—notice the `TIMEOUT: 20/10`, which is saying that when the 10 iterations of test were run, it timed out in 20 of those iterations. So that basically doesn’t make sense and I can vaguely recall the cause is a known issue in the test infrastructure that’s happened a lot of time before, but I don’t remember the cause. Anyway, that doesn’t need to block merging.

File: beacon/beacon-common.js
Comment (cdumez) [inline beacon/beacon-common.js]: Running this in WebKit & Blink, I see runTests() getting called twice (so we run all 7 sample tests twice, a total of 14 tests). This seems to indicate that the onload handler is called twice. Isn't it the expected behavior that unload gets called twice given that the iframe navigates itself to Microsoft.com at some point?

File: beacon/beacon-common.js
Comment (cdumez) [inline beacon/beacon-common.js]: Adding "iframe.onload = null;" here makes WebKit / Blink run only 7 tests as expected (runTests is only called once).
That said, all 7 tests time out in WebKit & Blink and I have not figured out why yet. Firefox seems to run all 7 tests fine though..

File: beacon/fetch-keepalive-navigate.iFrame.html
Comment (cdumez) [inline beacon/fetch-keepalive-navigate.iFrame.html]: Shouldn't this be .length instead of .Length ?

**Comment by toddreifsteck** [comment]
@igrigorik @Brandr0id Lets avoid taking new tests into this PR. After it is landed, new tests are "easy". Brandon will have enough on his hands to land this for us while juggling the other tasks he has on his plate.

**Comment by igrigorik** [comment]
@toddreifsteck sgtm.

File: beacon/beacon-common.js
Comment (cdumez) [inline beacon/beacon-common.js]: So one issue with Blink / WebKit is that you need to set the src attribute *before* adding the iframe to the document. Otherwise, onload gets called twice, once for about:blank load, and a second time for the fetch-keepalive-navigate.iFrame.html load.

File: beacon/beacon-common.js
Comment (cdumez) [inline beacon/beacon-common.js]: In WebKit / Blink, I need to do:
iframe.onload = null;

here, otherwise, the onload handler gets called again for the navigation to Microsoft.com at the end of the test and we run 14 tests instead of 7.

File: beacon/fetch-keepalive-navigate.iFrame.html
Comment (cdumez) [inline beacon/fetch-keepalive-navigate.iFrame.html]: Can we navigate to "about:blank" instead. Using this URL, I see this in the console:
Refused to display 'https://www.microsoft.com/en-us/' in a frame because it set 'X-Frame-Options' to 'sameorigin'.

**Comment by youennf** [comment]
Renaming beacon-common.js as beacon-common.sub.js might be nice.
This would remove the need for ?pipe=sub.

Looking at the polling mechanism, raising an error after some long enough time like 5 seconds would help limit the time needed to run the tests in case they time out.

File: beacon/beacon-common.js
Comment (Brandr0id) [inline beacon/beacon-common.js]: Swapped this around :)

File: beacon/beacon-common.js
Comment (Brandr0id) [inline beacon/beacon-common.js]: This may also be why the Chrome test pass is seeing 20/10 failures. Updated this.

File: beacon/fetch-keepalive-navigate.iFrame.html
Comment (Brandr0id) [inline beacon/fetch-keepalive-navigate.iFrame.html]: I updated it to use the root of the test server, does that work?

File: beacon/fetch-keepalive-navigate.iFrame.html
Comment (Brandr0id) [inline beacon/fetch-keepalive-navigate.iFrame.html]: indeed it should be, updated this.

File: beacon/beacon-cors.window.js
Comment (Brandr0id) [inline beacon/beacon-cors.window.js]: I agree with the other comments, let's land this PR then expand/add the extra test cases :)

**Comment by Brandr0id** [comment]
@jgraham \ @toddreifsteck This isn't really specific to this PR but I noticed the Edge run failing some tests due to behavior that was changed with the Creators Update. It looks like sauce labs is hard coded to use the anniversary update. Is there any particular reason this isn't either using the creators update '15.15063', or 'latest' which I think Sauce supports?

**Comment by Brandr0id** [comment]
Ping @igrigorik

File: beacon/beacon-basic-string.html
Comment (igrigorik) [inline beacon/beacon-basic-string.html]: @cdumez any findings on your end for this one?

**Review by igrigorik — APPROVED** [review]
As noted earlier, the CORS logic in https://github.com/w3c/web-platform-tests/pull/6655/files#diff-fcdc2e3ae8101103a5474d1a0f415b56 needs work as it doesn't cover the non-safelisted content type cases, but we'll handle that in a separate commit. 

Otherwise, LGTM. Thanks a lot for the great work on this!

**Pull request merged**
